core/src/main/java/org/geysermc/geyser/session/cache/TagCache.java

```diff
@@ -25,117 +25,139 @@
+                // Hack btw
+                boolean emulatePost1_13Logic = registryTags.get(MinecraftKey.key("signs")).length > 1;
+                session.setEmulatePost1_13Logic(emulatePost1_13Logic);
+                if (logger.isDebug()) {
+                    logger.debug("Emulating post 1.13 villager logic for " + session.bedrockUsername() + "? " + emulatePost1_13Logic);
                 }
-            } else {
-                localValues[tag.ordinal()] = IntArrays.EMPTY_ARRAY;
-                GeyserImpl.getInstance().getLogger().debug(type + " tag not found from server: " + location);
             }
-        });
+
+            Object2IntMap<Key> tagIndexMap = new Object2IntOpenHashMap<>();
+            this.tags[registry.ordinal()] = loadTags(registryTags, tagIndexMap, registry);
+            this.tagIndexMaps[registry.ordinal()] = tagIndexMap;
+        }
+    }
+
+    private int[][] loadTags(Map<Key, int[]> packetTags, Object2IntMap<Key> tagIndexMap, TagRegistry registry) {
+        Map<Key, Tag> vanillaTags = registry.getVanillaTags();
+        List<Key> nonVanillaTagKeys = packetTags.keySet().stream().filter(tag -> !vanillaTags.containsKey(tag)).toList();
+
+        int[][] tags = new int[vanillaTags.size() + nonVanillaTagKeys.size()][];
+
+        // Load all vanilla tags first (whether the server sent them or not), then load all the remaining non-vanilla tags the server might have sent.
+        for (Map.Entry<Key, Tag> vanillaTag : vanillaTags.entrySet()) {
+            tags[((VanillaTag) vanillaTag.getValue()).ordinal()] = packetTags.getOrDefault(vanillaTag.getKey(), new int[0]);
+        }
+
+        int tagIndex = vanillaTags.size();
+        for (Key nonVanillaTagKey : nonVanillaTagKeys) {
+            tags[tagIndex] = packetTags.get(nonVanillaTagKey);
+            tagIndexMap.put(nonVanillaTagKey, tagIndex);
+            tagIndex++;
+        }
+
+        return tags;
     }
 
     /**
      * @return true if the block tag is present and contains this block mapping's Java ID.
      */
-    public boolean is(BlockTag tag, Block block) {
-        int[] values = this.blocks[tag.ordinal()];
-        return contains(values, block.javaId());
+    public boolean is(Tag tag, Block block) {
+        assert tag.registry() == TagRegistry.BLOCK;
+        return contains(get(tag), block.javaId());
     }
 
     /**
      * @return true if the item tag is present and contains this item stack's Java ID.
      */
-    public boolean is(ItemTag tag, GeyserItemStack itemStack) {
+    public boolean is(Tag tag, GeyserItemStack itemStack) {
         return is(tag, itemStack.asItem());
     }
 
     /**
      * @return true if the item tag is present and contains this item's Java ID.
      */
-    public boolean is(ItemTag tag, Item item) {
-        int[] values = this.items[tag.ordinal()];
-        return contains(values, item.javaId());
+    public boolean is(Tag tag, Item item) {
+        assert tag.registry() == TagRegistry.ITEM;
```

Comment:
Would be careful about `assert` - this is a throwable and not an exception, so it can cause some havoc.

Reply:
Hm, should I throw an `IllegalArgumentException` instead or just put a debug log message?

Reply:
IllegalArgumentException should be fine.

Reply:
Used IllegalArgumentException in latest commit